pkg/webhook/utils.go

```diff
@@ -52,32 +52,37 @@ const (
 )
 
 // AddToManagerBuilder aggregates various AddToManager functions.
-type AddToManagerBuilder []func(manager.Manager) (webhook.Webhook, error)
+type AddToManagerBuilder map[string]func(manager.Manager) (webhook.Webhook, error)
 
 // NewAddToManagerBuilder creates a new AddToManagerBuilder and registers the given functions.
-func NewAddToManagerBuilder(funcs ...func(manager.Manager) (webhook.Webhook, error)) AddToManagerBuilder {
-	var builder AddToManagerBuilder
-	builder.Register(funcs...)
+func NewAddToManagerBuilder(params ...interface{}) AddToManagerBuilder {
+	builder := AddToManagerBuilder(make(map[string]func(manager.Manager) (webhook.Webhook, error)))
+	builder.Register(params...)
 	return builder
 }
 
 // Register registers the given functions in this builder.
-func (a *AddToManagerBuilder) Register(funcs ...func(manager.Manager) (webhook.Webhook, error)) {
-	*a = append(*a, funcs...)
+func (a AddToManagerBuilder) Register(params ...interface{}) {
+	for i := 0; i < len(params)-1; i += 2 {
+		name, f := params[i].(string), params[i+1].(func(manager.Manager) (webhook.Webhook, error))
+		a[name] = f
+	}
 }
 
 // AddToManager creates a webhook server adds all webhooks created by the AddToManager-functions of this builder to it.
 // It exits on the first error and returns it.
-func (a *AddToManagerBuilder) AddToManager(mgr manager.Manager, cfg *cmd.WebhookServerConfig) error {
+func (a *AddToManagerBuilder) AddToManager(mgr manager.Manager, cfg *cmd.WebhookServerConfig, disabled map[string]bool) error {
```

Comment:
Same as for the 'regular' `AddToManagerBuilder`, I'd not consider this the responsibility of the `AddToManagerBuilder` and move the logic to disable something outside.

Reply:
See above, please explain how exactly you think this could be done (and why it's better). IMO, `AddToManagerBuilder` is exactly the right abstraction to modify - it allows collecting all functions and then switching on / off which of them should be added based on a set of keys, exactly what I need.

Reply:
Same comment as above, an `AddToManagerBuilder` should not be involved with some logic like that - it doesn't even know about what each function does and thus cannot make decisions on what to add and what not - the right entity being able to decide what goes in is a level higher.

Reply:
I agree, and it isn't - the names to filter on are just passed from outside. The builder itself doesn't do any decisions at all.

Reply:
I would tend to disagree - in the new code, the builder does make decisions in my opinion:
https://github.com/gardener/gardener-extensions/pull/77/files/a3ddf255d0354ff81901eb09159581935bc3ff40#diff-ff368195ba700fad1507e2a15dfb8b09R79
Here you can see that the code and also the behavior of the builder is changed and it now decides on a map with which it actually does not have any relation.

Reply:
The builder filters out functions based on a filter passed from the outside. The builder is passive (it just uses the filter) and agnostic to where this filter comes from. Honestly, I don't see a problem with this approach.

Reply:
Problem is that the filter does not mean anything to the `AddToManagerBuilder`. It doesn't have any relation to the names, so why put some domain specific filtering logic in there? Also, as mentioned, with this approach it's possible to specify unknown controller names without causing an error although the tool should indeed error and fail early showing the user a misconfiguration.